Add flush to FileOutputStream and remove default no-op flush

ekrich opened the pull request:
When looking at this it didn't make sense so this seemed the right thing to do.

MasseGuillaume commented:
rebase from master to fix the ci: https://github.com/scala-native/scala-native/pull/704

ekrich commented:
@Duhemm Are you able to review this PR?

Duhemm commented:
@ekrich That's on top of my list! Thanks for looking at `FileOutputStream` 😉

Duhemm commented on javalib/src/main/scala/java/io/OutputStream.scala:
```diff
@@ -19,7 +19,7 @@ abstract class OutputStream extends Object with Closeable with Flushable {
   }
 
-  def flush(): Unit = ()
+  def flush(): Unit
```
`flush()` is not `abstract` in `OutputStream`

ekrich replied:
Thanks, I should read the javadoc.

Duhemm commented on javalib/src/main/scala/java/io/OutputStream.scala:
```diff
@@ -19,7 +19,7 @@ abstract class OutputStream extends Object with Closeable with Flushable {
+  def flush(): Unit
 
-  def close(): Unit = ()
+  def close(): Unit
```
`close()` is not `abstract` in `OutputStream`

ekrich replied:
Ditto.

Duhemm commented on javalib/src/main/scala/java/io/ByteArrayOutputStream.scala:
```diff
@@ -46,6 +46,8 @@ class ByteArrayOutputStream(initBufSize: Int) extends OutputStream {
   def toString(charsetName: String): String =
     new String(buf, 0, count, charsetName)
 
+  override def flush(): Unit = ()
```
This becomes unnecessary if `flush()` is not abstract.

ekrich replied:
Ditto.

ekrich commented:
It doesn't look like `FileOutputStream` overrides `flush` in harmony and I don't see how `readOnly` is handled exactly. I am correcting some problems to make it behave like the JVM version.

ekrich commented:
@Duhemm I have addressed the issues and made some changes to the behavior. Using `readOnly` avoids the call to `fsync` but we would have to have the native code modify the `FIleDescriptor.readOnly` field.

ekrich commented:
@Duhemm Do you think it is okay to add the readOnly as default false to the private constructor so I can make it immutable?

Duhemm commented:
@ekrich I think it's fine to add it to the constructor. The test can be a normal unit test. You should be able to use things like `java.io.File.createTempFile` for instance!

ekrich commented:
Good suggestion, thank you. I feel like I went down a rabbit hole on this one - teaching me a good lesson. Can't imagine what you have done with all this `java.io/nio` - it is a big job.

Duhemm commented on javalib/src/main/scala/java/io/FileDescriptor.scala:
```diff
@@ -3,17 +3,26 @@ package java.io
-    unistd.fsync(fd) match {
-      case 0 => ()
-      case _ => throw new java.io.SyncFailedException("sync failed")
+    if (fd <= 2) throwSyncFailed()
```
The POSIX spec says that `stdin`, `stdout` and `stderr` have value `0`, `1` and `2`, so that's ok 👍

Duhemm commented on javalib/src/main/scala/java/io/FileDescriptor.scala:
```diff
@@ -3,17 +3,26 @@ package java.io
     }
 
-  def valid(): Boolean = fd == -1
+  def valid(): Boolean = fd != -1
```
🤦‍♂️  Thanks!

ekrich replied:
Thanks for the tips and encouragement!
